File: Services/LeaveService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using TaskManagement.Data;
using TaskManagement.DTOs;

namespace TaskManagement.Services
{
    public interface ILeaveService
    {
        // Lets the "New Leave Request" form show the day count live as the user picks dates,
        // without duplicating the Holiday/WorkingDay day-type math on the frontend —
        // reuses the exact same ComputeDayCountAsync that RequestAsync/UpdateAsync save with,
        // so the preview can never drift from what actually gets persisted.
        Task<ApiResponse<decimal>> PreviewDayCountAsync(DateTime start, DateTime end);
        Task<ApiResponse<LeaveRequestDto>> RequestAsync(int userId, CreateLeaveRequestDto dto);
        Task<ApiResponse<LeaveRequestDto>> UpdateAsync(int userId, int requestId, UpdateLeaveRequestDto dto);
        Task<ApiResponse<List<LeaveRequestDto>>> GetMyRequestsAsync(int userId, int page = 1, int pageSize = 25);
        Task<ApiResponse<List<LeaveRequestDto>>> GetAllRequestsAsync(int? filterUserId, string? filterStatus, int page = 1, int pageSize = 25);
        Task<ApiResponse<LeaveRequestDto>> DecideAsync(int approverId, int requestId, DecideLeaveRequestDto dto);
        Task<ApiResponse<bool>> DeleteAsync(int userId, int requestId);
        Task<ApiResponse<LeaveRequestDto>> SetPermissionsAsync(int requestId, SetLeaveRequestPermissionsDto dto);
        Task<ApiResponse<LeaveBalanceDto>> GetMyBalanceAsync(int userId);
    }

    public class LeaveService : ILeaveService
    {
        // Sane cap on a single request's span so a fat-fingered range doesn't silently succeed.
        private const int MaxRequestSpanDays = 90;

        private readonly PMSDbContext _context;
        private readonly IWorkweekRulesService _rules;
        private readonly IAnnualLeaveAllocationService _allocations;

        public LeaveService(PMSDbContext context, IWorkweekRulesService rules, IAnnualLeaveAllocationService allocations)
        {
            _context = context;
            _rules = rules;
            _allocations = allocations;
        }

        // Sum of "day contribution" (0 / 1) for each date in [start, end] inclusive, driven by
        // IWorkweekRulesService.GetDayTypeAsync so an off Saturday and any Holiday-table row
        // (including admin-added named holidays) are both accounted for.
        private async Task<decimal> ComputeDayCountAsync(DateTime start, DateTime end)
        {
            decimal total = 0;
            for (var d = start.Date; d <= end.Date; d = d.AddDays(1))
            {
                var dayType = await _rules.GetDayTypeAsync(d);
                total += dayType == "Holiday" ? 0m : 1m;
            }
            return total;
        }

        // Shared by RequestAsync (create) and UpdateAsync (edit) — excludeRequestId lets an edit's
        // overlap check ignore the very request being edited.
        private async Task<(bool Ok, string? Error, decimal DayCount, LeaveType? LeaveType)> ValidateAndComputeAsync(
            int userId, int leaveTypeId, DateTime start, DateTime end, int? excludeRequestId)
        {
            if ((end - start).TotalDays > MaxRequestSpanDays)
                return (false, $"Leave range cannot exceed {MaxRequestSpanDays} days.", 0, null);

            var today = AppClock.Today;
            if (start < today)
                return (false, "Leave cannot be requested for a past date — only today or a future date is allowed.", 0, null);

            var leaveType = await _context.LeaveTypes.FindAsync(leaveTypeId);
            if (leaveType == null)
                return (false, "Invalid leave type.", 0, null);

            var overlapQuery = _context.LeaveRequests.Where(r =>
                r.UserId == userId &&
                (r.Status == "Pending" || r.Status == "Approved") &&
                r.StartDate <= end && r.EndDate >= start);
            if (excludeRequestId.HasValue) overlapQuery = overlapQuery.Where(r => r.Id != excludeRequestId.Value);
            if (await overlapQuery.AnyAsync())
                return (false, "This range overlaps an existing pending or approved leave request.", 0, null);

            var dayCount = await ComputeDayCountAsync(start, end);
            if (dayCount <= 0)
                return (false, "The selected range contains no working days.", 0, null);

            // No balance gate here by design — a user may still submit a request after their
            // annual allocation is fully used (or already exceeded); the balance is informational
            // (surfaced via GetMyBalanceAsync) and it's the approver's call whether to approve a
            // request that runs over.
            return (true, null, dayCount, leaveType);
        }

        public async Task<ApiResponse<decimal>> PreviewDayCountAsync(DateTime start, DateTime end)
        {
            var s = start.Date;
            var e = end.Date;

            if (e < s)
                return new ApiResponse<decimal> { Success = false, Message = "End date must be on or after the start date." };
            if ((e - s).TotalDays > MaxRequestSpanDays)
                return new ApiResponse<decimal> { Success = false, Message = $"Leave range cannot exceed {MaxRequestSpanDays} days." };

            return new ApiResponse<decimal> { Success = true, Data = await ComputeDayCountAsync(s, e) };
        }

        public async Task<ApiResponse<LeaveRequestDto>> RequestAsync(int userId, CreateLeaveRequestDto dto)
        {
            // ValidationFilter/CreateLeaveRequestDtoValidator already guarantee StartDate/EndDate
            // are non-null and EndDate >= StartDate by the time this runs.
            var start = dto.StartDate!.Value.Date;
            var end = dto.EndDate!.Value.Date;

            var (ok, error, dayCount, leaveType) = await ValidateAndComputeAsync(userId, dto.LeaveTypeId, start, end, excludeRequestId: null);
            if (!ok) return new ApiResponse<LeaveRequestDto> { Success = false, Message = error };

            var request = new LeaveRequest
            {
                UserId = userId,
                LeaveTypeId = leaveType!.Id,
                StartDate = start,
                EndDate = end,
                DayCount = dayCount,
                Reason = dto.Reason.Trim(),
                Status = "Pending",
                CreatedAt = AppClock.Now,
            };
            _context.LeaveRequests.Add(request);
            await _context.SaveChangesAsync();

            await _context.Entry(request).Reference(r => r.User).LoadAsync();
            request.LeaveType = leaveType;

            return new ApiResponse<LeaveRequestDto> { Success = true, Data = ToDto(request) };
        }

        public async Task<ApiResponse<LeaveRequestDto>> UpdateAsync(int userId, int requestId, UpdateLeaveRequestDto dto)
        {
            var request = await _context.LeaveRequests
                .Include(r => r.User).Include(r => r.LeaveType)
                .FirstOrDefaultAsync(r => r.Id == requestId && r.UserId == userId);
            if (request == null) return ApiResponse<LeaveRequestDto>.NotFound("Leave request not found.");

            if (request.Status != "Pending")
                return new ApiResponse<LeaveRequestDto> { Success = false, Message = "Only pending requests can be edited." };
            if (!request.AllowEdit)
                return new ApiResponse<LeaveRequestDto> { Success = false, Message = "Editing has been disabled for this request by an admin." };

            var start = dto.StartDate!.Value.Date;
            var end = dto.EndDate!.Value.Date;

            var (ok, error, dayCount, leaveType) = await ValidateAndComputeAsync(userId, dto.LeaveTypeId, start, end, excludeRequestId: requestId);
            if (!ok) return new ApiResponse<LeaveRequestDto> { Success = false, Message = error };

            request.LeaveTypeId = leaveType!.Id;
            request.LeaveType = leaveType;
            request.StartDate = start;
            request.EndDate = end;
            request.DayCount = dayCount;
            request.Reason = dto.Reason.Trim();
            await _context.SaveChangesAsync();

            return new ApiResponse<LeaveRequestDto> { Success = true, Data = ToDto(request) };
        }

        public async Task<ApiResponse<List<LeaveRequestDto>>> GetMyRequestsAsync(int userId, int page = 1, int pageSize = 25)
        {
            pageSize = Math.Clamp(pageSize, 1, 100);
            page = Math.Max(1, page);

            var query = _context.LeaveRequests
                .Include(r => r.User).Include(r => r.LeaveType).Include(r => r.Approver)
                .Where(r => r.UserId == userId)
                .OrderByDescending(r => r.CreatedAt);

            var totalCount = await query.CountAsync();
            var totalPages = totalCount == 0 ? 1 : (int)Math.Ceiling((double)totalCount / pageSize);

            var list = await query
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();

            return new ApiResponse<List<LeaveRequestDto>>
            {
                Success = true,
                Data = list.Select(ToDto).ToList(),
                TotalCount = totalCount,
                Page = page,
                PageSize = pageSize,
                TotalPages = totalPages
            };
        }

        public async Task<ApiResponse<List<LeaveRequestDto>>> GetAllRequestsAsync(int? filterUserId, string? filterStatus, int page = 1, int pageSize = 25)
        {
            pageSize = Math.Clamp(pageSize, 1, 100);
            page = Math.Max(1, page);

            var query = _context.LeaveRequests
                .Include(r => r.User).Include(r => r.LeaveType).Include(r => r.Approver)
                .AsQueryable();

            if (filterUserId.HasValue) query = query.Where(r => r.UserId == filterUserId.Value);
            if (!string.IsNullOrWhiteSpace(filterStatus)) query = query.Where(r => r.Status == filterStatus);

            var totalCount = await query.CountAsync();
            var totalPages = totalCount == 0 ? 1 : (int)Math.Ceiling((double)totalCount / pageSize);

            var list = await query
                .OrderByDescending(r => r.CreatedAt)
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();

            return new ApiResponse<List<LeaveRequestDto>>
            {
                Success = true,
                Data = list.Select(ToDto).ToList(),
                TotalCount = totalCount,
                Page = page,
                PageSize = pageSize,
                TotalPages = totalPages
            };
        }

        public async Task<ApiResponse<LeaveRequestDto>> DecideAsync(int approverId, int requestId, DecideLeaveRequestDto dto)
        {
            var request = await _context.LeaveRequests
                .Include(r => r.User).Include(r => r.LeaveType)
                .FirstOrDefaultAsync(r => r.Id == requestId);
            if (request == null) return ApiResponse<LeaveRequestDto>.NotFound("Leave request not found.");

            if (request.Status != "Pending")
                return new ApiResponse<LeaveRequestDto> { Success = false, Message = $"This request has already been {request.Status.ToLower()}." };

            request.Status = dto.Approve ? "Approved" : "Rejected";
            request.ApproverId = approverId;
            request.DecisionAt = AppClock.Now;
            request.DecisionNote = dto.DecisionNote?.Trim();
            await _context.SaveChangesAsync();

            await _context.Entry(request).Reference(r => r.Approver).LoadAsync();

            return new ApiResponse<LeaveRequestDto> { Success = true, Data = ToDto(request) };
        }

        public async Task<ApiResponse<bool>> DeleteAsync(int userId, int requestId)
        {
            var request = await _context.LeaveRequests.FirstOrDefaultAsync(r => r.Id == requestId && r.UserId == userId);
            if (request == null) return ApiResponse<bool>.NotFound("Leave request not found.");

            if (request.Status != "Pending")
                return new ApiResponse<bool> { Success = false, Message = "Only pending requests can be deleted." };
            if (!request.AllowDelete)
                return new ApiResponse<bool> { Success = false, Message = "Deleting has been disabled for this request by an admin." };

            _context.LeaveRequests.Remove(request);
            await _context.SaveChangesAsync();

            return new ApiResponse<bool> { Success = true, Data = true };
        }

        public async Task<ApiResponse<LeaveRequestDto>> SetPermissionsAsync(int requestId, SetLeaveRequestPermissionsDto dto)
        {
            var request = await _context.LeaveRequests
                .Include(r => r.User).Include(r => r.LeaveType).Include(r => r.Approver)
                .FirstOrDefaultAsync(r => r.Id == requestId);
            if (request == null) return ApiResponse<LeaveRequestDto>.NotFound("Leave request not found.");

            if (request.Status != "Pending")
                return new ApiResponse<LeaveRequestDto> { Success = false, Message = "Edit/delete permissions can only be changed while a request is pending." };

            request.AllowEdit = dto.AllowEdit;
            request.AllowDelete = dto.AllowDelete;
            await _context.SaveChangesAsync();

            return new ApiResponse<LeaveRequestDto> { Success = true, Data = ToDto(request) };
        }

        public async Task<ApiResponse<LeaveBalanceDto>> GetMyBalanceAsync(int userId)
        {
            var year = AppClock.Today.Year;
            var allocated = await _allocations.GetLeaveDaysForYearAsync(year);
            var used = await GetUsedDaysAsync(userId, year);

            return new ApiResponse<LeaveBalanceDto>
            {
                Success = true,
                Data = new LeaveBalanceDto
                {
                    Year = year,
                    AllocatedDays = allocated,
                    UsedDays = used,
                    AvailableDays = allocated - used,
                }
            };
        }

        // Sum of Approved requests' DayCount for a user in a given year, across every leave type
        // (the pooled allocation doesn't distinguish by type).
        private async Task<decimal> GetUsedDaysAsync(int userId, int year, int? excludeRequestId = null)
        {
            var query = _context.LeaveRequests
                .Where(r => r.UserId == userId && r.Status == "Approved" && r.StartDate.Year == year);
            if (excludeRequestId.HasValue) query = query.Where(r => r.Id != excludeRequestId.Value);
            return await query.SumAsync(r => (decimal?)r.DayCount) ?? 0m;
        }

        private static LeaveRequestDto ToDto(LeaveRequest r) => new()
        {
            Id = r.Id,
            UserId = r.UserId,
            UserFullName = r.User?.FullName ?? string.Empty,
            UserAvatarUrl = r.User?.AvatarUrl,
            LeaveTypeId = r.LeaveTypeId,
            LeaveTypeName = r.LeaveType?.Name ?? string.Empty,
            StartDate = r.StartDate,
            EndDate = r.EndDate,
            DayCount = r.DayCount,
            Reason = r.Reason,
            Status = r.Status,
            ApproverId = r.ApproverId,
            ApproverName = r.Approver?.FullName,
            DecisionAt = r.DecisionAt,
            DecisionNote = r.DecisionNote,
            AllowEdit = r.AllowEdit,
            AllowDelete = r.AllowDelete,
            CreatedAt = r.CreatedAt,
        };
    }
}

File: ClientApp/src/services/leave.service.ts
import { apiRequest, apiRequestWithMeta, PagedResult } from '../lib/api';

export interface PaginatedResponse<T> {
  data: T[];
  totalCount: number;
  page: number;
  pageSize: number;
  totalPages: number;
}

export type LeaveRequestStatus = 'Pending' | 'Approved' | 'Rejected';
export type DayType = 'Holiday' | 'WorkingDay';

export interface LeaveRequest {
  id: number;
  userId: number;
  userFullName: string;
  userAvatarUrl?: string;
  leaveTypeId: number;
  leaveTypeName: string;
  startDate: string;
  endDate: string;
  dayCount: number;
  reason: string;
  status: LeaveRequestStatus;
  approverId?: number;
  approverName?: string;
  decisionAt?: string;
  decisionNote?: string;
  // Admin-controlled, meaningful only while Pending — gate the owner's Edit/Delete buttons.
  allowEdit: boolean;
  allowDelete: boolean;
  createdAt: string;
}

export interface CreateLeaveRequestDto {
  leaveTypeId: number;
  startDate: string;
  endDate: string;
  reason: string;
}

export type UpdateLeaveRequestDto = CreateLeaveRequestDto;

export interface DecideLeaveRequestDto {
  approve: boolean;
  decisionNote?: string;
}

export interface SetLeaveRequestPermissionsDto {
  allowEdit: boolean;
  allowDelete: boolean;
}

export interface LeaveType {
  id: number;
  name: string;
  isActive: boolean;
}

export interface SaveLeaveTypeDto {
  name: string;
  isActive: boolean;
}

// Type-agnostic — one pooled balance per user per year, sourced from AnnualLeaveAllocation.
export interface LeaveBalance {
  year: number;
  allocatedDays: number;
  usedDays: number;
  availableDays: number;
}

export interface AnnualLeaveAllocation {
  year: number;
  leaveDays: number;
}

export interface UpdateAnnualLeaveAllocationDto {
  leaveDays: number;
}

export interface Holiday {
  id: number;
  date: string;
  name: string;
  dayType: DayType;
  isManualOverride: boolean;
}

export interface SetHolidayOverrideDto {
  date: string;
  name: string;
  dayType: DayType;
}

export interface WorkweekRules {
  workStartTime: string;
  workEndTime: string;
  breakMinMinutes: number;
  breakMaxMinutes: number;
  holidaySaturdayOccurrences: number[];
  updatedAt?: string;
}

export interface UpdateWorkweekRulesDto {
  workStartTime: string;
  workEndTime: string;
  breakMinMinutes: number;
  breakMaxMinutes: number;
  holidaySaturdayOccurrences: number[];
}

export const leaveService = {
  // ── Requests ──────────────────────────────────────────────────────────
  async getMyRequests(page = 1, pageSize = 25): Promise<PaginatedResponse<LeaveRequest>> {
    const params = new URLSearchParams({ page: String(page), pageSize: String(pageSize) });
    const result = await apiRequestWithMeta<LeaveRequest[]>(`/leave/requests/mine?${params.toString()}`);
    return {
      data: result.data,
      totalCount: result.meta.totalCount,
      page: result.meta.page,
      pageSize: result.meta.pageSize,
      totalPages: result.meta.totalPages,
    };
  },

  // Server-computed day count for a not-yet-submitted date range (Holiday/WorkingDay
  // math lives once in LeaveService — this just previews it live instead of duplicating it
  // client-side). Silent so it doesn't flash the global loader as dates are picked.
  async previewDayCount(start: string, end: string): Promise<number> {
    const p = new URLSearchParams({ start, end });
    return apiRequest<number>(`/leave/requests/preview?${p.toString()}`, {}, { silent: true });
  },

  async getAllRequests(params?: { userId?: number; status?: string }, page = 1, pageSize = 25): Promise<PaginatedResponse<LeaveRequest>> {
    const p = new URLSearchParams({ page: String(page), pageSize: String(pageSize) });
    if (params?.userId != null) p.set('userId', String(params.userId));
    if (params?.status) p.set('status', params.status);
    const result = await apiRequestWithMeta<LeaveRequest[]>(`/leave/requests/all?${p.toString()}`);
    return {
      data: result.data,
      totalCount: result.meta.totalCount,
      page: result.meta.page,
      pageSize: result.meta.pageSize,
      totalPages: result.meta.totalPages,
    };
  },

  async createRequest(dto: CreateLeaveRequestDto): Promise<LeaveRequest> {
    return apiRequest<LeaveRequest>('/leave/requests', {
      method: 'POST',
      body: JSON.stringify(dto),
    });
  },

  async updateRequest(id: number, dto: UpdateLeaveRequestDto): Promise<LeaveRequest> {
    return apiRequest<LeaveRequest>(`/leave/requests/${id}`, {
      method: 'PUT',
      body: JSON.stringify(dto),
    });
  },

  async decide(id: number, dto: DecideLeaveRequestDto): Promise<LeaveRequest> {
    return apiRequest<LeaveRequest>(`/leave/requests/${id}/decide`, {
      method: 'POST',
      body: JSON.stringify(dto),
    });
  },

  async setPermissions(id: number, dto: SetLeaveRequestPermissionsDto): Promise<LeaveRequest> {
    return apiRequest<LeaveRequest>(`/leave/requests/${id}/permissions`, {
      method: 'PUT',
      body: JSON.stringify(dto),
    });
  },

  async deleteRequest(id: number): Promise<boolean> {
    return apiRequest<boolean>(`/leave/requests/${id}`, { method: 'DELETE' });
  },

  // ── Balances ──────────────────────────────────────────────────────────
  async getMyBalance(): Promise<LeaveBalance> {
    return apiRequest<LeaveBalance>('/leave/balances/mine');
  },

  // ── Leave types (admin CRUD) ─────────────────────────────────────────
  async getTypes(): Promise<LeaveType[]> {
    return apiRequest<LeaveType[]>('/leave/types');
  },

  // For dropdowns - uses search endpoint with pagination
  async searchTypes(page = 1, pageSize = 25, search?: string): Promise<PaginatedResponse<LeaveType>> {
    const params = new URLSearchParams({ page: String(page), pageSize: String(pageSize) });
    if (search) params.set('q', search);
    
    const result = await apiRequestWithMeta<LeaveType[]>(`/leave/types/search?${params.toString()}`);
    return {
      data: result.data,
      totalCount: result.meta.totalCount,
      page: result.meta.page,
      pageSize: result.meta.pageSize,
      totalPages: result.meta.totalPages,
    };
  },

  async createType(dto: SaveLeaveTypeDto): Promise<LeaveType> {
    return apiRequest<LeaveType>('/leave/types', { method: 'POST', body: JSON.stringify(dto) });
  },

  async updateType(id: number, dto: SaveLeaveTypeDto): Promise<LeaveType> {
    return apiRequest<LeaveType>(`/leave/types/${id}`, { method: 'PUT', body: JSON.stringify(dto) });
  },

  async deleteType(id: number): Promise<boolean> {
    return apiRequest<boolean>(`/leave/types/${id}`, { method: 'DELETE' });
  },

  // ── Holidays ──────────────────────────────────────────────────────────
  async getHolidays(from: string, to: string): Promise<Holiday[]> {
    const p = new URLSearchParams({ from, to });
    return apiRequest<Holiday[]>(`/leave/holidays?${p.toString()}`);
  },

  async setHolidayOverride(dto: SetHolidayOverrideDto): Promise<Holiday> {
    return apiRequest<Holiday>('/leave/holidays/override', { method: 'POST', body: JSON.stringify(dto) });
  },

  async deleteHoliday(id: number): Promise<boolean> {
    return apiRequest<boolean>(`/leave/holidays/${id}`, { method: 'DELETE' });
  },

  // ── Annual Leave Allocation (admin) ──────────────────────────────────
  async getAllocations(): Promise<AnnualLeaveAllocation[]> {
    return apiRequest<AnnualLeaveAllocation[]>('/leave/allocations');
  },

  async updateCurrentAllocation(dto: UpdateAnnualLeaveAllocationDto): Promise<AnnualLeaveAllocation> {
    return apiRequest<AnnualLeaveAllocation>('/leave/allocations/current', { method: 'PUT', body: JSON.stringify(dto) });
  },

  // ── Rules Settings ────────────────────────────────────────────────────
  async getRules(): Promise<WorkweekRules> {
    return apiRequest<WorkweekRules>('/leave/rules');
  },

  async updateRules(dto: UpdateWorkweekRulesDto): Promise<WorkweekRules> {
    return apiRequest<WorkweekRules>('/leave/rules', { method: 'PUT', body: JSON.stringify(dto) });
  },
};
